Support UPI Custom Model Deployment (#393)
caraml-dev/merlin: docs/user-guide/custom_model.md

# Custom Model

> Custom model enables users to to deploy any docker image that satisfy merlin requirements. Users is responsible to develop their own web service, build and publish the docker image, which later on can be deployed through Merlin.

Reviewer: Typo, should be
```
Custom model enables users to deploy any docker image that satisfy merlin requirements. Users are responsible to develop their own web service, build and publish the docker image, which later on can be deployed through Merlin.
```

## Why Custom Model 

> Users should consider to use custom model, if they have one of the following conditions:

Reviewer: The first reason is that if they want to use other languages than Python.

### UPI_V1 Custom Model
Similar with `HTTP_JSON` custom model, user can add the artifact during model uploading, and the uploaded artifacts will be available in the directory specified by `CARAML_ARTIFACT_LOCATION` environment variable. The web server must implement service that defined in the [UPI interface](https://github.com/caraml-dev/universal-prediction-interface/blob/main/proto/caraml/upi/v1/upi.proto#L11), also the server must open and listen to the port number given by `CARAML_GRPC_PORT` environment variable.

If users want to emit metrics from this web server, they need to create scrape metrics REST endpoint. The challenge here, the knative (the underlying k8s deployment tools that merlin use) doesn't open multiple ports, hence the REST endpoint must be running on the same port as gRPC server (using port number given by `CARAML_GRPC_PORT`). Not every programming language can support running multiple protocol (gRPC and HTTP in this case) on the same port, for Go language users can use [cmux](https://github.com/soheilhy/cmux) to solve this problem, otherwise users can use push metrics to [pushgateway](https://prometheus.io/docs/instrumenting/pushing/)

## Docker Image

> Docker image must contains web service application and dependencies that must be installed in order to run the web service. Users are responsible for building the docker image as well as for publishing it. Currently there are only two kind of docker registries that can be used:

Reviewer: We can remove this part since it's Gojek-specific
```
 Currently there are only two kind of docker registries that can be used:
* Artifactory
* GCR that is owned by gods GCP project
```

* Artifactory
* GCR that is owned by gods GCP project

## Deployment

Using Merlin SDK

```
resource_request = ResourceRequest(1, 1, "1", "1Gi")

> model_dir = "model_dir"

Reviewer: Add comment here to let user knows what can `model_dir` be used for.

with merlin.new_model_version() as v:
    v.log_custom_model(image="ghcr.io/yourcustommodelimage", model_dir=model_dir)

endpoint = merlin.deploy(v, resource_request= resource_request, protocol = Protocol.HTTP_JSON)
# endpoint = merlin.deploy(v, resource_request= resource_request, protocol = Protocol.UPI_V1) if using UPI

```

> ### Deployment Flow:

Reviewer: Can we add documentation of all environment variables that merlin passes to the model?

If users want to emit metrics from this web server, they need to create scrape metrics REST endpoint. The challenge here, the knative (the underlying k8s deployment tools that merlin use) doesn't open multiple ports, hence the REST endpoint must be running on the same port as gRPC server (using port number given by `CARAML_GRPC_PORT`). Not every programming language can support running multiple protocol (gRPC and HTTP in this case) on the same port, for Go language users can use [cmux](https://github.com/soheilhy/cmux) to solve this problem, otherwise users can use push metrics to [pushgateway](https://prometheus.io/docs/instrumenting/pushing/)

### Environment Variables
As mentioned in the previous section, there are several environment variables that will be supplied by merlin control plan to custom model. Below are the list of the variables

| Name | Description |
|------|-------------|
| STORAGE_URI | Contains the URI where the `model` artifacts is remotely stored |

> | CARAML_HTTP_PORT | Port that must be listen and open by custom model server that use `HTTP_JSON` protocol |

Reviewer: ```suggestion
| CARAML_HTTP_PORT | Port that must be openend when the model is deployed with `HTTP_JSON` protocol |
```

> | CARAML_GRPC_PORT | Port that must be listen and open by custom model server that use `UPI_V1` protocol |

Reviewer: ```suggestion
| CARAML_GRPC_PORT | Port that must be openend when the model is deployed with `UPI_V1` protocol |
```
